File: api_sw/app/Module/Service/Auth/Role.php
<?php

declare(strict_types=1);

namespace App\Module\Service\Auth;

use App\Module\Db\Dao\Auth\ActionRelToScene;
use App\Module\Db\Dao\Auth\Menu;
use App\Module\Db\Dao\Auth\RoleRelOfPlatformAdmin;
use App\Module\Db\Dao\Auth\RoleRelToAction;
use App\Module\Db\Dao\Auth\RoleRelToMenu;
use App\Module\Logic\Auth\Role as AuthRole;
use App\Module\Service\AbstractService;

class Role extends AbstractService
{
    /**
     * 创建
     *
     * @param array $data
     * @return void
     */
    public function create(array $data)
    {
        if (isset($data['menuIdArr']) && count($data['menuIdArr']) != getDao(Menu::class)->parseFilter(['id' => $data['menuIdArr'], 'sceneId' => $data['sceneId']])->getBuilder()->count()) {
            //$count = getDao(Menu::class)->parseFilter(['id' => $data['menuIdArr'], 'sceneId' => $data['sceneId'] ?? $oldInfo->sceneId])->info();
            throwFailJson(89999998);
        }
        if (isset($data['actionIdArr']) && count($data['actionIdArr']) != getDao(ActionRelToScene::class)->parseFilter(['actionId' => $data['actionIdArr'], 'sceneId' => $data['sceneId']])->getBuilder()->count()) {
            throwFailJson(89999998);
        }

        $id = $this->getDao()->parseInsert($data)->insert();
        if (empty($id)) {
            throwFailJson();
        }
        if (isset($data['menuIdArr'])) {
            $this->container->get(AuthRole::class)->saveRelMenu($data['menuIdArr'], $id);
        }
        if (isset($data['actionIdArr'])) {
            $this->container->get(AuthRole::class)->saveRelAction($data['actionIdArr'], $id);
        }
        throwSuccessJson();
    }

    /**
     * 更新
     *
     * @param array $data
     * @param array $filter
     * @return void
     */
    public function update(array $data, array $filter)
    {
        if (isset($data['menuIdArr']) || isset($data['actionIdArr'])) {
            $idArr = $this->getIdArr($filter);
            foreach ($idArr as $id) {
                $filterOne = ['roleId'=>$id];
                $oldInfo = $this->getDao()->parseFilter($filterOne)->info();
                $this->getDao()->parseFilter($filterOne)->parseUpdate($data)->update();    //有可能只改menuIdArr或actionIdArr
                if (isset($data['menuIdArr'])) {
                    if (count($data['menuIdArr']) != getDao(Menu::class)->parseFilter(['id' => $data['menuIdArr'], 'sceneId' => $data['sceneId'] ?? $oldInfo->sceneId])->getBuilder()->count()) {
                        throwFailJson(89999998);
                    }
                    $this->container->get(AuthRole::class)->saveRelMenu($data['menuIdArr'], $oldInfo->roleId);
                }
                if (isset($data['actionIdArr'])) {
                    if (count($data['actionIdArr']) != getDao(ActionRelToScene::class)->parseFilter(['actionId' => $data['actionIdArr'], 'sceneId' => $data['sceneId'] ?? $oldInfo->sceneId])->getBuilder()->count()) {
                        throwFailJson(89999998);
                    }
                    $this->container->get(AuthRole::class)->saveRelAction($data['actionIdArr'], $oldInfo->roleId);
                }
            }
        } else {
            $result = $this->getDao()->parseFilter($filter)->parseUpdate($data)->update();
            if (empty($result)) {
                throwFailJson();
            }
        }
        throwSuccessJson();
    }

    /**
     * 删除
     *
     * @param array $filter
     * @return void
     */
    public function delete(array $filter)
    {
        $idArr = $this->getIdArr($filter);
        $result = $this->getDao()->parseFilter($filter)->delete();
        if (empty($result)) {
            throwFailJson();
        }
        getDao(RoleRelToMenu::class)->parseFilter(['roleId' => $idArr])->delete();
        getDao(RoleRelToAction::class)->parseFilter(['roleId' => $idArr])->delete();
        getDao(RoleRelOfPlatformAdmin::class)->parseFilter(['roleId' => $idArr])->delete();
        throwSuccessJson();
    }
}

File: api/internal/logic/auth/role.go
package auth

import (
	daoAuth "api/internal/dao/auth"
	"api/internal/service"
	"api/internal/utils"
	"context"

	"github.com/gogf/gf/v2/database/gdb"
	"github.com/gogf/gf/v2/frame/g"
	"github.com/gogf/gf/v2/util/gconv"
)

type sAuthRole struct{}

func NewAuthRole() *sAuthRole {
	return &sAuthRole{}
}

func init() {
	service.RegisterAuthRole(NewAuthRole())
}

// 验证数据（create和update共用）
func (logicThis *sAuthRole) verifyData(ctx context.Context, data map[string]any) (err error) {
	if _, ok := data[daoAuth.Role.Columns().SceneId]; ok && gconv.String(data[daoAuth.Role.Columns().SceneId]) != `` {
		if count, _ := daoAuth.Scene.CtxDaoModel(ctx).FilterPri(data[daoAuth.Role.Columns().SceneId]).Count(); count == 0 {
			err = utils.NewErrorCode(ctx, 29999997, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.scene`)}})
			return
		}
	}

	/* if _, ok := data[`action_id_arr`]; ok && len(gconv.Strings(data[`action_id_arr`])) > 0 {
		actionIdArr := gconv.Strings(data[`action_id_arr`])
		if count, _ := daoAuth.Action.CtxDaoModel(ctx).FilterPri(actionIdArr).Count(); count != len(actionIdArr) {
			err = utils.NewErrorCode(ctx, 29999997, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.action`)}})
			return
		}
	}

	if _, ok := data[`menu_id_arr`]; ok && len(gconv.Uints(data[`menu_id_arr`])) > 0 {
		menuIdArr := gconv.Uints(data[`menu_id_arr`])
		if count, _ := daoAuth.Menu.CtxDaoModel(ctx).FilterPri(menuIdArr).Count(); count != len(menuIdArr) {
			err = utils.NewErrorCode(ctx, 29999997, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.menu`)}})
			return
		}
	} */
	return
}

// 新增
func (logicThis *sAuthRole) Create(ctx context.Context, data map[string]any) (id any, err error) {
	if err = logicThis.verifyData(ctx, data); err != nil {
		return
	}
	daoModelThis := daoAuth.Role.CtxDaoModel(ctx)

	if _, ok := data[`action_id_arr`]; ok && len(gconv.Strings(data[`action_id_arr`])) > 0 {
		actionIdArr := gconv.Strings(data[`action_id_arr`])
		if count, _ := daoAuth.ActionRelToScene.CtxDaoModel(ctx).Filters(g.Map{daoAuth.ActionRelToScene.Columns().ActionId: actionIdArr, daoAuth.ActionRelToScene.Columns().SceneId: data[daoAuth.Role.Columns().SceneId]}).Count(); count != len(actionIdArr) {
			err = utils.NewErrorCode(ctx, 29999997, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.action`)}})
			return
		}
	}

	if _, ok := data[`menu_id_arr`]; ok && len(gconv.Uints(data[`menu_id_arr`])) > 0 {
		menuIdArr := gconv.Uints(data[`menu_id_arr`])
		if count, _ := daoAuth.Menu.CtxDaoModel(ctx).Filters(g.Map{daoAuth.Menu.Columns().MenuId: menuIdArr, daoAuth.Menu.Columns().SceneId: data[daoAuth.Role.Columns().SceneId]}).Count(); count != len(menuIdArr) {
			err = utils.NewErrorCode(ctx, 29999997, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.menu`)}})
			return
		}
	}

	id, err = daoModelThis.HookInsert(data).InsertAndGetId()
	return
}

// 修改
func (logicThis *sAuthRole) Update(ctx context.Context, filter map[string]any, data map[string]any) (row int64, err error) {
	if err = logicThis.verifyData(ctx, data); err != nil {
		return
	}
	daoModelThis := daoAuth.Role.CtxDaoModel(ctx)

	sceneIdArr := []string{}
	sceneId := ``
	var list gdb.Result
	if _, ok := data[`action_id_arr`]; ok && len(gconv.Strings(data[`action_id_arr`])) > 0 {
		if len(sceneIdArr) == 0 {
			if _, ok := data[daoAuth.Role.Columns().SceneId]; ok {
				sceneIdArr = append(sceneIdArr, gconv.String(data[daoAuth.Role.Columns().SceneId]))
			} else {
				if len(list) == 0 {
					list, _ = daoAuth.Role.CtxDaoModel(ctx).Filters(filter).Fields(daoAuth.Role.Columns().RelId, daoAuth.Role.Columns().SceneId).All()
					if len(list) == 0 {
						err = utils.NewErrorCode(ctx, 29999998, ``)
						return
					}
				}
				sceneIdSet := map[string]struct{}{}
				for _, info := range list {
					sceneId := info[daoAuth.Role.Columns().SceneId].String()
					if _, ok := sceneIdSet[sceneId]; !ok {
						sceneIdSet[sceneId] = struct{}{}
						sceneIdArr = append(sceneIdArr, sceneId)
					}
				}
				/* if len(sceneIdSet) != 1 {
					err = utils.NewErrorCode(ctx, 89999998, ``)
					return
				} */
			}
		}
		actionIdArr := gconv.Strings(data[`action_id_arr`])
		for _, sceneId := range sceneIdArr {
			if count, _ := daoAuth.ActionRelToScene.CtxDaoModel(ctx).Filters(g.Map{daoAuth.ActionRelToScene.Columns().SceneId: sceneId, daoAuth.ActionRelToScene.Columns().ActionId: actionIdArr}).Count(); count != len(actionIdArr) {
				err = utils.NewErrorCode(ctx, 29999997, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.action`)}})
				return
			}
		}
	}

	if _, ok := data[`menu_id_arr`]; ok && len(gconv.Uints(data[`menu_id_arr`])) > 0 {
		if sceneId == `` {
			if _, ok := data[daoAuth.Role.Columns().SceneId]; ok {
				sceneId = gconv.String(data[daoAuth.Role.Columns().SceneId])
			} else {
				if len(list) == 0 {
					list, _ = daoAuth.Role.CtxDaoModel(ctx).Filters(filter).Fields(daoAuth.Role.Columns().RelId, daoAuth.Role.Columns().SceneId).All()
					if len(list) == 0 {
						err = utils.NewErrorCode(ctx, 29999998, ``)
						return
					}
				}
				sceneIdSet := map[string]struct{}{}
				for _, info := range list {
					sceneIdSet[info[daoAuth.Role.Columns().SceneId].String()] = struct{}{}
				}
				if len(sceneIdSet) != 1 { //菜单所属场景只能设置一个，故必须同一场景下的角色才能一起修改
					err = utils.NewErrorCode(ctx, 89999998, ``)
					return
				}
				sceneId = list[0][daoAuth.Role.Columns().SceneId].String()
			}
		}
		menuIdArr := gconv.Uints(data[`menu_id_arr`])
		if count, _ := daoAuth.Menu.CtxDaoModel(ctx).Filters(g.Map{daoAuth.Menu.Columns().SceneId: sceneId, daoAuth.Menu.Columns().MenuId: menuIdArr}).Count(); count != len(menuIdArr) {
			err = utils.NewErrorCode(ctx, 29999997, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.menu`)}})
			return
		}
	}

	if len(list) > 0 {
		idArr := []uint{}
		for _, info := range list {
			idArr = append(idArr, info[daoAuth.Role.Columns().RelId].Uint())
		}
		filter = map[string]any{`id`: idArr}
	}

	daoModelThis.SetIdArr(filter)
	if len(daoModelThis.IdArr) == 0 {
		err = utils.NewErrorCode(ctx, 29999998, ``)
		return
	}

	row, err = daoModelThis.HookUpdate(data).UpdateAndGetAffected()
	return
}

// 删除
func (logicThis *sAuthRole) Delete(ctx context.Context, filter map[string]any) (row int64, err error) {
	daoModelThis := daoAuth.Role.CtxDaoModel(ctx)

	daoModelThis.SetIdArr(filter)
	if len(daoModelThis.IdArr) == 0 {
		err = utils.NewErrorCode(ctx, 29999998, ``)
		return
	}

	if count, _ := daoAuth.RoleRelOfAdmin.CtxDaoModel(ctx).Filter(daoAuth.RoleRelOfAdmin.Columns().RoleId, daoModelThis.IdArr).Count(); count > 0 {
		err = utils.NewErrorCode(ctx, 30009999, ``, g.Map{`i18nValues`: []any{g.I18n().T(ctx, `name.auth.role`), count, g.I18n().T(ctx, `name.auth.roleRelOfAdmin`)}})
		return
	}

	row, err = daoModelThis.HookDelete().DeleteAndGetAffected()
	return
}
